From: =?utf-8?q?Ond=C5=99ej_Sur=C3=BD?= <ondrej@sury.org>
Date: Wed, 27 Jul 2016 11:04:16 +0200
Subject: Initialize error in tests/gd2/gd2_read.c

---
 tests/gd2/gd2_read.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/tests/gd2/gd2_read.c b/tests/gd2/gd2_read.c
index 94fe069..ce91df1 100644
--- a/tests/gd2/gd2_read.c
+++ b/tests/gd2/gd2_read.c
@@ -5,7 +5,7 @@
 
 int main(int argc, char *argv[])
 {
-	int error, i = 0;
+	int error = 0, i = 0;
 	gdImagePtr im, exp;
 	FILE *fp;
 	char *path[] = {
