diff -ruN libgd2-2.0.33/gd.c.orig libgd2-2.0.33/gd.c
--- libgd2-2.0.33/gd.c.orig	2006-04-21 11:31:18.000000000 +0800
+++ libgd2-2.0.33/gd.c	2006-04-21 11:31:15.000000000 +0800
@@ -3104,7 +3104,8 @@
 		/* TBB: set the last pixel for consistency (<=) */
 		while ((x >> 16) <= x2) {
 			gdImageSetAAPixelColor(im, x >> 16, y >> 16, col, (y >> 8) & 0xFF);
-			gdImageSetAAPixelColor(im, x >> 16, (y >> 16) + 1,col, (~y >> 8) & 0xFF);
+         if ((y >> 16) + 1 <= im->cy2)
+            gdImageSetAAPixelColor(im, x >> 16, (y >> 16) + 1,col, (~y >> 8) & 0xFF);
 			x += (1 << 16);
 			y += inc;
 		}
@@ -3125,7 +3126,8 @@
 		/* TBB: set the last pixel for consistency (<=) */
 		while ((y>>16) <= y2) {
 			gdImageSetAAPixelColor(im, x >> 16, y >> 16, col, (x >> 8) & 0xFF);
-			gdImageSetAAPixelColor(im, (x >> 16) + 1, (y >> 16),col, (~x >> 8) & 0xFF);
+         if ((x >> 16) + 1 <= im->cx2)
+            gdImageSetAAPixelColor(im, (x >> 16) + 1, (y >> 16),col, (~x >> 8) & 0xFF);
 			x += inc;
 			y += (1<<16);
 		}

