diff -ruN libgd2-2.0.33/gd.c.orig libgd2-2.0.33/gd.c
--- libgd2-2.0.33/gd.c	2006-04-21 16:37:55.000000000 +0800
+++ libgd2-2.0.33/gd.c 2006-04-21 16:39:07.000000000 +0800
@@ -3064,6 +3064,9 @@
 	im->tpixels[y][x]=gdTrueColorAlpha(dr, dg, db,  gdAlphaOpaque);
 }  
 
+/* simple helper */
+inline int min_int(int a, int b) { return (a < b ? a : b); }
+
 static void gdImageAALine (gdImagePtr im, int x1, int y1, int x2, int y2, int col)
 {
 	/* keep them as 32bits */
@@ -3074,11 +3077,15 @@
 		gdImageLine(im, x1, y1, x2, y2, col);
 		return;
 	}
-        /* TBB: use the clipping rectangle */
-        if (clip_1d (&x1, &y1, &x2, &y2, im->cx1, im->cx2) == 0)
-          return;
-        if (clip_1d (&y1, &x1, &y2, &x2, im->cy1, im->cy2) == 0)
-          return;
+
+  /* TBB: use the clipping rectangle */
+   /* use expanded image bounds to ensure we get all the AA pixels drawn on the edges */
+   /* Note the +1 and -1 */
+  if (clip_1d (&x1, &y1, &x2, &y2, im->cx1-1, im->cx2+1) == 0)
+     return;
+  if (clip_1d (&y1, &x1, &y2, &x2, im->cy1-1, im->cy2+1) == 0) 
+     return;
+
 	dx = x2 - x1;
 	dy = y2 - y1;
 
@@ -3098,15 +3105,25 @@
 			dx = x2 - x1;
 			dy = y2 - y1;
 		}
-		x = x1 << 16;
+		x = x1;
 		y = y1 << 16;
 		inc = (dy * 65536) / dx;
+
 		/* TBB: set the last pixel for consistency (<=) */
-		while ((x >> 16) <= x2) {
-			gdImageSetAAPixelColor(im, x >> 16, y >> 16, col, (y >> 8) & 0xFF);
-         if ((y >> 16) + 1 <= im->cy2)
-            gdImageSetAAPixelColor(im, x >> 16, (y >> 16) + 1,col, (~y >> 8) & 0xFF);
-			x += (1 << 16);
+      /* Constrain the loop to the image bounds */
+      if (x < 0)
+      {
+         y -= inc * x;
+         x = 0;
+      }
+      x2 = min_int(x2,im->cx2);
+		while (x <= x2) {
+         int py = y >> 16;
+         if (py >= im->cy1 && py <= im->cy2)
+            gdImageSetAAPixelColor(im, x, py, col, (y >> 8) & 0xFF);
+         if (py + 1 <= im->cy2)
+            gdImageSetAAPixelColor(im, x, py + 1,col, (~y >> 8) & 0xFF);
+			x += 1;
 			y += inc;
 		}
 	} else {
@@ -3121,15 +3138,26 @@
 			dy = y2 - y1;
 		}
 		x = x1 << 16;
-		y = y1 << 16;
+		// y = y1 << 16;
+      y = y1;
 		inc = (dx * 65536) / dy;
+
 		/* TBB: set the last pixel for consistency (<=) */
-		while ((y>>16) <= y2) {
-			gdImageSetAAPixelColor(im, x >> 16, y >> 16, col, (x >> 8) & 0xFF);
-         if ((x >> 16) + 1 <= im->cx2)
-            gdImageSetAAPixelColor(im, (x >> 16) + 1, (y >> 16),col, (~x >> 8) & 0xFF);
+      /* Constrain the loop to the image bounds */
+      if (y < 0)
+      {
+         x -= inc * y;
+         y = 0;
+      }
+      y2 = min_int(y2,im->cy2);
+		while (y <= y2) {
+         int px = x >> 16;
+         if (px >= im->cx1 && px <= im->cx2)
+            gdImageSetAAPixelColor(im, px, y, col, (x >> 8) & 0xFF);
+         if (px + 1 <= im->cx2)
+            gdImageSetAAPixelColor(im, px + 1, y,col, (~x >> 8) & 0xFF);
 			x += inc;
-			y += (1<<16);
+			y += 1;
 		}
 	}
 }

