diff -ruNp libgd2-2.0.33.orig/bdftogd libgd2-2.0.33/bdftogd
--- libgd2-2.0.33.orig/bdftogd	2004-10-27 16:22:04.000000000 +0200
+++ libgd2-2.0.33/bdftogd	2006-07-17 04:23:15.000000000 +0200
@@ -139,6 +139,10 @@ print FILEC <<"EOF";
 
 $info
 
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
 #include "$filename.h"
 
 char ${gdname}Data[] = {
diff -ruNp libgd2-2.0.33.orig/entities.h libgd2-2.0.33/entities.h
--- libgd2-2.0.33.orig/entities.h	2004-10-27 16:22:46.000000000 +0200
+++ libgd2-2.0.33/entities.h	2006-07-17 04:20:47.000000000 +0200
@@ -11,7 +11,7 @@
 extern "C" {
 #endif
 
-struct entities_s {
+static struct entities_s {
 	char	*name;
 	int	value;
 } entities[] = {
diff -ruNp libgd2-2.0.33.orig/entities.tcl libgd2-2.0.33/entities.tcl
--- libgd2-2.0.33.orig/entities.tcl	2004-10-27 16:22:48.000000000 +0200
+++ libgd2-2.0.33/entities.tcl	2006-07-17 04:20:47.000000000 +0200
@@ -32,7 +32,7 @@ puts $f "#ifdef __cplusplus"
 puts $f "extern \"C\" {"
 puts $f "#endif"
 puts $f ""
-puts $f "struct entities_s {"
+puts $f "static struct entities_s {"
 puts $f "	char	*name;"
 puts $f "	int	value;"
 puts $f "} entities\[\] = {"
diff -ruNp libgd2-2.0.33.orig/gd.c libgd2-2.0.33/gd.c
--- libgd2-2.0.33.orig/gd.c	2006-07-17 04:20:19.000000000 +0200
+++ libgd2-2.0.33/gd.c	2006-07-17 04:20:47.000000000 +0200
@@ -73,6 +73,8 @@ BGD_DECLARE(gdImagePtr) gdImageCreate (i
   int i;
   gdImagePtr im;
   im = (gdImage *) gdMalloc (sizeof (gdImage));
+  if (!im)
+    return NULL;
   memset (im, 0, sizeof (gdImage));
   /* Row-major ever since gd 1.3 */
   if (sy >= INT_MAX/sizeof (unsigned char *) ||
@@ -89,6 +91,15 @@ BGD_DECLARE(gdImagePtr) gdImageCreate (i
     {
       /* Row-major ever since gd 1.3 */
       im->pixels[i] = (unsigned char *) gdCalloc (sx, sizeof (unsigned char));
+      if (!im->pixels[i]) 
+	{
+	  for (--i ; i >= 0; i--)
+	    {
+	      gdFree(im->pixels[i]);
+	    }
+	  gdFree(im);
+	  return NULL;
+	}
     }
   im->sx = sx;
   im->sy = sy;
diff -ruNp libgd2-2.0.33.orig/gdfontg.c libgd2-2.0.33/gdfontg.c
--- libgd2-2.0.33.orig/gdfontg.c	2004-10-27 16:22:04.000000000 +0200
+++ libgd2-2.0.33/gdfontg.c	2006-07-17 04:20:47.000000000 +0200
@@ -10,6 +10,9 @@
    "Libor Skarvada, libor@informatics.muni.cz"
  */
 
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
 
 #include "gdfontg.h"
 
diff -ruNp libgd2-2.0.33.orig/gdfontl.c libgd2-2.0.33/gdfontl.c
--- libgd2-2.0.33.orig/gdfontl.c	2004-10-27 16:22:04.000000000 +0200
+++ libgd2-2.0.33/gdfontl.c	2006-07-17 04:20:47.000000000 +0200
@@ -11,6 +11,9 @@
    "Libor Skarvada, libor@informatics.muni.cz"
  */
 
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
 
 #include "gdfontl.h"
 
diff -ruNp libgd2-2.0.33.orig/gdfontmb.c libgd2-2.0.33/gdfontmb.c
--- libgd2-2.0.33.orig/gdfontmb.c	2004-10-27 16:22:04.000000000 +0200
+++ libgd2-2.0.33/gdfontmb.c	2006-07-17 04:20:47.000000000 +0200
@@ -9,6 +9,9 @@
    No copyright info was found in the original bdf.
  */
 
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
 
 #include "gdfontmb.h"
 
diff -ruNp libgd2-2.0.33.orig/gdfonts.c libgd2-2.0.33/gdfonts.c
--- libgd2-2.0.33.orig/gdfonts.c	2004-10-27 16:22:04.000000000 +0200
+++ libgd2-2.0.33/gdfonts.c	2006-07-17 04:20:47.000000000 +0200
@@ -7,6 +7,9 @@
    No copyright info was found in the original bdf.
  */
 
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
 
 #include "gdfonts.h"
 
diff -ruNp libgd2-2.0.33.orig/gdfontt.c libgd2-2.0.33/gdfontt.c
--- libgd2-2.0.33.orig/gdfontt.c	2004-10-27 16:22:04.000000000 +0200
+++ libgd2-2.0.33/gdfontt.c	2006-07-17 04:20:47.000000000 +0200
@@ -10,6 +10,9 @@
    "Libor Skarvada, libor@informatics.muni.cz"
  */
 
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
 
 #include "gdfontt.h"
 
diff -ruNp libgd2-2.0.33.orig/gdft.c libgd2-2.0.33/gdft.c
--- libgd2-2.0.33.orig/gdft.c	2004-11-02 22:00:48.000000000 +0100
+++ libgd2-2.0.33/gdft.c	2006-07-17 04:20:47.000000000 +0200
@@ -832,7 +832,7 @@ BGD_DECLARE(char *) gdImageStringFTEx (g
   FT_Matrix matrix;
   FT_Vector penf, oldpenf, delta, total_min = {0,0}, total_max = {0,0}, glyph_min, glyph_max;
   FT_Face face;
-  FT_CharMap charmap;
+  FT_CharMap charmap = NULL;
   FT_Glyph image;
   FT_GlyphSlot slot;
   FT_Error err;
@@ -945,6 +945,9 @@ BGD_DECLARE(char *) gdImageStringFTEx (g
   /* set rotation transform */
   FT_Set_Transform (face, &matrix, NULL);
 
+#if 0
+fprintf(stderr, "fontsize=%g resolution=%d,%d\n", ptsize, METRIC_RES, METRIC_RES );
+#endif
   FT_New_Size (face, &platform_independent);
   FT_Activate_Size (platform_independent);
   if (FT_Set_Char_Size (face, 0, (FT_F26Dot6)(ptsize*64), METRIC_RES, METRIC_RES))
@@ -956,18 +959,23 @@ BGD_DECLARE(char *) gdImageStringFTEx (g
 
   if (render)
     {
+#if 0
+fprintf(stderr, "fontsize=%g dpi=%d,%d\n", ptsize, hdpi, vdpi );
+#endif
       FT_New_Size (face, &platform_specific);
       FT_Activate_Size (platform_specific);
       if (FT_Set_Char_Size (face, 0, (FT_F26Dot6)(ptsize*64), hdpi, vdpi))
         {
           gdCacheDelete (tc_cache);
           gdMutexUnlock (gdFontCacheMutex);
-          return "Could not set character size";
+          return "Could not set character size (render)";
         }
     }
 
-  if (fg < 0)
-      render_mode |= FT_LOAD_MONOCHROME;
+#if 0
+    if (fg < 0)
+	render_mode |= FT_LOAD_MONOCHROME;
+#endif
 
   /* find requested charmap */
   encodingfound = 0;
@@ -1208,12 +1216,14 @@ fprintf(stderr,"dpi=%d,%d metric_res=%d 
 		so we don't have to recheck for the terminating number */
 	  if (! xshow_alloc) {
 		xshow_alloc = 100;
-		strex->xshow = malloc(xshow_alloc);
+		strex->xshow = gdMalloc(xshow_alloc);
 		xshow_pos = 0;
 	  } 
 	  else if (xshow_pos + 20 > xshow_alloc) {
 		xshow_alloc += 100;
 		strex->xshow = realloc(strex->xshow, xshow_alloc);
+		if (! strex->xshow)
+	  	    return "Problem allocating memory with realloc";
 	}
 	  xshow_pos += sprintf(strex->xshow + xshow_pos, "%g ",
 		(double)(penf.x - oldpenf.x) * hdpi / (64 * METRIC_RES));
@@ -1322,10 +1332,8 @@ fprintf(stderr,"dpi=%d,%d metric_res=%d 
 
   if (brect)
     {				/* only if need brect */
-      double dpix, dpiy;
-      
-      dpix = 64 * METRIC_RES / hdpi;
-      dpiy = 64 * METRIC_RES / vdpi;
+      double scalex = (double)hdpi / (64 * METRIC_RES);
+      double scaley = (double)vdpi / (64 * METRIC_RES);
 
       /* increase by 1 pixel to allow for rounding */
       total_min.x -= METRIC_RES;
@@ -1334,14 +1342,14 @@ fprintf(stderr,"dpi=%d,%d metric_res=%d 
       total_max.y += METRIC_RES;
  
       /* rotate bounding rectangle, scale and round to int pixels, and translate */
-      brect[0] = x + (total_min.x * cos_a + total_max.y * sin_a)/dpix;
-      brect[1] = y - (total_min.x * sin_a - total_max.y * cos_a)/dpiy;
-      brect[2] = x + (total_max.x * cos_a + total_max.y * sin_a)/dpix;
-      brect[3] = y - (total_max.x * sin_a - total_max.y * cos_a)/dpiy;
-      brect[4] = x + (total_max.x * cos_a + total_min.y * sin_a)/dpix;
-      brect[5] = y - (total_max.x * sin_a - total_min.y * cos_a)/dpiy;
-      brect[6] = x + (total_min.x * cos_a + total_min.y * sin_a)/dpix;
-      brect[7] = y - (total_min.x * sin_a - total_min.y * cos_a)/dpiy;
+      brect[0] = x + (total_min.x * cos_a + total_max.y * sin_a)*scalex;
+      brect[1] = y - (total_min.x * sin_a - total_max.y * cos_a)*scaley;
+      brect[2] = x + (total_max.x * cos_a + total_max.y * sin_a)*scalex;
+      brect[3] = y - (total_max.x * sin_a - total_max.y * cos_a)*scaley;
+      brect[4] = x + (total_max.x * cos_a + total_min.y * sin_a)*scalex;
+      brect[5] = y - (total_max.x * sin_a - total_min.y * cos_a)*scaley;
+      brect[6] = x + (total_min.x * cos_a + total_min.y * sin_a)*scalex;
+      brect[7] = y - (total_min.x * sin_a - total_min.y * cos_a)*scaley;
     }
 
   FT_Done_Size (platform_independent);
@@ -1510,6 +1518,66 @@ static char * font_pattern(char **fontpa
 
 #endif /* HAVE_LIBFONTCONFIG */
 
+#ifdef WIN32
+#define FONTSFX "\\FONTS"
+#define GDP     "GDFONTPATH="
+#define F_OK 0			/* Needed in Windows */
+
+/* chkWinFontPath:
+ * See if environment variable envv is defined and
+ * if the path $envv/FONTS exists. If so, return the path,
+ * else return NULL.
+ */
+static char*
+chkWinFontPath (char* envv)
+{
+	char* root;
+	char* path;
+	root = getenv(envv);
+	if (!root) return NULL;
+	path = (char*)gdMalloc(strlen(root)+strlen(FONTSFX)+1);
+	strcpy (path, root);
+	strcat (path, FONTSFX);
+	if (access (path, F_OK) == 0) { /* success */
+		return path;
+	}
+	else {
+		gdFree (path);
+		return NULL;
+	}
+}
+
+/* getWinFontPath:
+ * Find Windows system font directory. Use environment variables to
+ * find system directory with a FONTS subdirectory.
+ * If successful, set GDFONTPATH to this value, for caching, and return it.
+ * Else, return NULL
+ */
+static char*
+getWinFontPath ()
+{
+	char* path = chkWinFontPath("SystemRoot");
+	if (!path) {
+		path = chkWinFontPath ("WinDir");
+		if (!path)
+			return NULL;
+	}
+#ifdef HAVE_SETENV
+	setenv("GDFONTPATH", path, 1);
+#else
+	{
+		char* ebuf = (char*)gdMalloc(strlen(GDP)+strlen(path)+1);
+		int rv;
+		strcpy(ebuf,GDP);
+		strcat(ebuf,path);
+		rv = putenv(ebuf);
+	}
+#endif
+	return path;
+}
+
+#endif  /* WIN32 */
+
 /* Look up font using font names as file names. */
 static char * font_path(char **fontpath, char *name_list)
 {
@@ -1524,6 +1592,10 @@ static char * font_path(char **fontpath,
    */
   *fontpath = NULL;
   fontsearchpath = getenv ("GDFONTPATH");
+#ifdef WIN32
+  if (!fontsearchpath)
+    fontsearchpath = getWinFontPath ();
+#endif
   if (!fontsearchpath)
     fontsearchpath = DEFAULT_FONTPATH;
   fontlist = strdup (name_list);
diff -ruNp libgd2-2.0.33.orig/gdfx.c libgd2-2.0.33/gdfx.c
--- libgd2-2.0.33.orig/gdfx.c	2004-10-27 16:22:07.000000000 +0200
+++ libgd2-2.0.33/gdfx.c	2006-07-17 04:20:47.000000000 +0200
@@ -1,3 +1,7 @@
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
 #include "gd.h"
 #include <math.h>
 
@@ -228,16 +232,26 @@ main (int argc, char *argv[])
   /* Create an image of text on a circle, with an
      alpha channel so that we can copy it onto a
      background */
+
+  im = NULL;
+
+/* What is this eleanor.jpg stuff doing here anyway?
+ * There is no such file in the gd-2.0.33 sources. */
+#if 0
+#if HAVE_LIBJPEG
   in = fopen ("eleanor.jpg", "rb");
-  if (!in)
-    {
-      im = gdImageCreateTrueColor (300, 300);
-    }
-  else
+  if (in)
     {
       im = gdImageCreateFromJpeg (in);
       fclose (in);
     }
+#endif
+#endif
+
+  if (!im)
+    {
+      im = gdImageCreateTrueColor (300, 300);
+    }
   if (gdImageSX (im) < gdImageSY (im))
     {
       radius = gdImageSX (im) / 2;
diff -ruNp libgd2-2.0.33.orig/gd_gif_in.c libgd2-2.0.33/gd_gif_in.c
--- libgd2-2.0.33.orig/gd_gif_in.c	2006-07-17 04:20:12.000000000 +0200
+++ libgd2-2.0.33/gd_gif_in.c	2006-07-17 04:20:47.000000000 +0200
@@ -1,3 +1,7 @@
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
 #include <stdio.h>
 #include <math.h>
 #include <string.h>
diff -ruNp libgd2-2.0.33.orig/gd_gif_out.c libgd2-2.0.33/gd_gif_out.c
--- libgd2-2.0.33.orig/gd_gif_out.c	2004-11-02 14:41:00.000000000 +0100
+++ libgd2-2.0.33/gd_gif_out.c	2006-07-17 04:20:47.000000000 +0200
@@ -1,3 +1,7 @@
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
 #include <stdio.h>
 #include <math.h>
 #include <string.h>
diff -ruNp libgd2-2.0.33.orig/gdhelpers.h libgd2-2.0.33/gdhelpers.h
--- libgd2-2.0.33.orig/gdhelpers.h	2004-10-28 19:09:12.000000000 +0200
+++ libgd2-2.0.33/gdhelpers.h	2006-07-17 04:20:47.000000000 +0200
@@ -29,30 +29,38 @@ int overflow2(int a, int b);
 
 /* 2.0.16: portable mutex support for thread safety. */
 
-#ifdef WIN32
-/* 2.0.18: must include windows.h to get CRITICAL_SECTION. */
-#include <windows.h>
-#define gdMutexDeclare(x) CRITICAL_SECTION x
-#define gdMutexSetup(x) InitializeCriticalSection(&x)
-#define gdMutexShutdown(x) DeleteCriticalSection(&x)
-#define gdMutexLock(x) EnterCriticalSection(&x)
-#define gdMutexUnlock(x) LeaveCriticalSection(&x)
-#else
-#ifdef HAVE_PTHREAD
-#include <pthread.h>
-#define gdMutexDeclare(x) pthread_mutex_t x
-#define gdMutexSetup(x) pthread_mutex_init(&x, 0)
-#define gdMutexShutdown(x) pthread_mutex_destroy(&x)
-#define gdMutexLock(x) pthread_mutex_lock(&x)
-#define gdMutexUnlock(x) pthread_mutex_unlock(&x)
+#ifdef DISABLE_THREADS
+#  define gdMutexDeclare(x)
+#  define gdMutexSetup(x)
+#  define gdMutexShutdown(x)
+#  define gdMutexLock(x)
+#  define gdMutexUnlock(x)
 #else
-#define gdMutexDeclare(x)
-#define gdMutexSetup(x) 
-#define gdMutexShutdown(x) 
-#define gdMutexLock(x) 
-#define gdMutexUnlock(x) 
-#endif /* HAVE_PTHREAD */
-#endif /* WIN32 */
+#  ifdef WIN32
+/* 2.0.18: must include windows.h to get CRITICAL_SECTION. */
+#    include <windows.h>
+#    define gdMutexDeclare(x) CRITICAL_SECTION x
+#    define gdMutexSetup(x) InitializeCriticalSection(&x)
+#    define gdMutexShutdown(x) DeleteCriticalSection(&x)
+#    define gdMutexLock(x) EnterCriticalSection(&x)
+#    define gdMutexUnlock(x) LeaveCriticalSection(&x)
+#  else
+#    ifdef HAVE_PTHREAD
+#      include <pthread.h>
+#      define gdMutexDeclare(x) pthread_mutex_t x
+#      define gdMutexSetup(x) pthread_mutex_init(&x, 0)
+#      define gdMutexShutdown(x) pthread_mutex_destroy(&x)
+#      define gdMutexLock(x) pthread_mutex_lock(&x)
+#      define gdMutexUnlock(x) pthread_mutex_unlock(&x)
+#    else
+#      define gdMutexDeclare(x)
+#      define gdMutexSetup(x) 
+#      define gdMutexShutdown(x) 
+#      define gdMutexLock(x) 
+#      define gdMutexUnlock(x) 
+#    endif /* HAVE_PTHREAD */
+#  endif /* WIN32 */
+#endif /* DISABLE_THREADS */
 
 #endif /* GDHELPERS_H */
 
diff -ruNp libgd2-2.0.33.orig/gdparttopng.c libgd2-2.0.33/gdparttopng.c
--- libgd2-2.0.33.orig/gdparttopng.c	2004-10-27 16:22:04.000000000 +0200
+++ libgd2-2.0.33/gdparttopng.c	2006-07-17 04:20:47.000000000 +0200
@@ -39,7 +39,11 @@ main (int argc, char **argv)
 
   printf ("Extracting from (%d, %d), size is %dx%d\n", x, y, w, h);
 
+#ifdef HAVE_LIBZ
   im = gdImageCreateFromGd2Part (in, x, y, w, h);
+#else
+  im = NULL;
+#endif
   fclose (in);
   if (!im)
     {
diff -ruNp libgd2-2.0.33.orig/gd_topal.c libgd2-2.0.33/gd_topal.c
--- libgd2-2.0.33.orig/gd_topal.c	2004-10-28 20:12:08.000000000 +0200
+++ libgd2-2.0.33/gd_topal.c	2006-07-17 04:20:47.000000000 +0200
@@ -49,6 +49,10 @@
   #define ORIGINAL_LIB_JPEG_REVERSE_ODD_ROWS
  */
 
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
 #include <string.h>
 #include "gd.h"
 #include "gdhelpers.h"
