Description: Add missing pointer dereference in font_pattern
Author: Colin Watson <cjwatson@ubuntu.com>
Bug-Debian: http://bugs.debian.org/709050
Last-Update: 2013-06-01

--- libgd2.orig/src/gdft.c
+++ libgd2/src/gdft.c
@@ -1487,7 +1487,7 @@ static char * font_pattern(char **fontpa
 			return "could not alloc font path";
 		}
 		strncpy(*fontpath, (const char *)file, file_len);
-		fontpath[file_len] = 0;
+		(*fontpath)[file_len] = 0;
 	}
 	FcPatternDestroy(font);
 
